runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -401,7 +402,14 @@ private void setTimer(WindowedValue<InputT> timerElement, TimerInternals.TimerDa
   @SuppressWarnings("ByteBufferBackingArray")
   private void fireCleanupTimers() {
     while (!cleanupTimers.isEmpty()) {
-      InternalTimer<?, TimerInternals.TimerData> timer = cleanupTimers.remove();
+      KV<RemoteBundle, InternalTimer<?, TimerInternals.TimerData>> kv = cleanupTimers.peek();
+      if (kv.getKey() != null && kv.getKey() == sdkHarnessRunner.remoteBundle) {
+        // user timer and cleanup may trigger together in finish bundle,
+        // defer state cleanup to allow user timers to complete
+        return;
```

Comment:
Wouldn't it be easier to fire timers only after finishing the bundle? That way we would guarantee for user timers to be fired.

Reply:
This gets executed during finish bundle and will ensure that the cleanup timers fire after the bundle only.

Reply:
Let's move it to where it belongs, i.e. after the bundle has been finished.